### ISSUE TITLE
Slow indexing

### ISSUE THREAD

**Issue by Edouard360** [YosefLab/scVI-dev#63]
Code has become very slow probably due to an indexing mistake on our side... @maxime1310 and I will try to identify the bottleneck

**Comment by jeff-regier** [comment]
Sounds good. This might be a good time to get familiar with the profiler. #43

**Comment by Edouard360** [comment]
So I first ran on my CPU locally:
```
def loop_over_dataset():
    gene_dataset_train = CortexDataset('train')
    dataloader = DataLoader(gene_dataset_train, batch_size=32)
    tic = time.time()
    for i in range(10):
        for _ in dataloader:
            pass
    toc = time.time()
    print(toc - tic)
```
Numpy dense: 2.16 s.
Scipy Sparse : 0.27 s.

And then I tried on AWS GPU with brain_large dataset (sampled, as in Romain's preprocessing, with 20k), this time, **timing the full training procedure** over 30 epochs instead of just the loop over the dataloader :

Numpy dense: 482 s.
Scipy Sparse : 479 s.

So this difference is probably not a bottleneck, and we might further evidence it analyzing the process profiling (I just left a snakeviz screenshot below):

<img width="1408" alt="screen shot 2018-04-17 at 21 27 18" src="https://user-images.githubusercontent.com/15527397/38911796-73b09c1e-4286-11e8-88d6-eae55c651c40.png">

The more detailed output, looking only at the dataloader lines is the following

<div id="pstats-table_wrapper" class="dataTables_wrapper no-footer"><div id="pstats-table_filter" class="dataTables_filter"></div><table cellpadding="0" cellspacing="0" border="0" class="display dataTable no-footer" id="pstats-table" role="grid" aria-describedby="pstats-table_info" style="width: 1369px;">
<thead><tr role="row"><th class="sorting" tabindex="0" aria-controls="pstats-table" rowspan="1" colspan="1" aria-label="ncalls: activate to sort column ascending" style="width: 135px;">ncalls</th><th class="sorting" tabindex="0" aria-controls="pstats-table" rowspan="1" colspan="1" aria-label="tottime: activate to sort column ascending" style="width: 84px;">tottime</th><th class="sorting" tabindex="0" aria-controls="pstats-table" rowspan="1" colspan="1" aria-label="percall: activate to sort column ascending" style="width: 84px;">percall</th><th class="sorting" tabindex="0" aria-controls="pstats-table" rowspan="1" colspan="1" aria-label="cumtime: activate to sort column ascending" style="width: 100px;">cumtime</th><th class="sorting_desc" tabindex="0" aria-controls="pstats-table" rowspan="1" colspan="1" aria-label="percall: activate to sort column ascending" style="width: 85px;" aria-sort="descending">percall</th><th class="sorting" tabindex="0" aria-controls="pstats-table" rowspan="1" colspan="1" aria-label="filename:lineno(function): activate to sort column ascending" style="width: 665px;">filename:lineno(function)</th></tr></thead><tbody><tr role="row" class="odd"><td>12</td><td class="">0.000177</td><td>1.475e-05</td><td>0.2421</td><td class="sorting_1">0.02017</td><td>dataloader.py:416(__iter__)</td></tr><tr role="row" class="even"><td>12</td><td class="">0.002392</td><td>0.0001993</td><td>0.2419</td><td class="sorting_1">0.02016</td><td>dataloader.py:186(__init__)</td></tr><tr role="row" class="odd"><td>1896</td><td class="">0.0235</td><td>1.239e-05</td><td>3.765</td><td class="sorting_1">0.001986</td><td>dataloader.py:256(__next__)</td></tr><tr role="row" class="even"><td>1884</td><td class="">0.008584</td><td>4.556e-06</td><td>2.487</td><td class="sorting_1">0.00132</td><td>dataloader.py:247(_get_batch)</td></tr><tr role="row" class="odd"><td>1908</td><td class="">0.01552</td><td>8.134e-06</td><td>1.271</td><td class="sorting_1">0.0006659</td><td>dataloader.py:288(_put_indices)</td></tr><tr role="row" class="even"><td>1884</td><td class="">0.0084</td><td>4.459e-06</td><td>1.252</td><td class="sorting_1">0.0006644</td><td>dataloader.py:297(_process_next_batch)</td></tr><tr role="row" class="odd"><td>1</td><td class="">2.6e-05</td><td>2.6e-05</td><td>0.000632</td><td class="sorting_1">0.000632</td><td>dataloader.py:1(&lt;module&gt;)</td></tr><tr role="row" class="even"><td>24</td><td class="">0.000327</td><td>1.362e-05</td><td>0.002855</td><td class="sorting_1">0.000119</td><td>dataloader.py:312(_shutdown_workers)</td></tr><tr role="row" class="odd"><td>12</td><td class="">0.000501</td><td>4.175e-05</td><td>0.000781</td><td class="sorting_1">6.508e-05</td><td>dataloader.py:158(_set_SIGCHLD_handler)</td></tr><tr role="row" class="even"><td>12</td><td class="">0.00013</td><td>1.083e-05</td><td>0.000651</td><td class="sorting_1">5.425e-05</td><td>dataloader.py:210(&lt;listcomp&gt;)</td></tr><tr role="row" class="odd"><td>11</td><td class="">0.000154</td><td>1.4e-05</td><td>0.000212</td><td class="sorting_1">1.927e-05</td><td>dataloader.py:172(handler)</td></tr><tr role="row" class="even"><td>2</td><td class="">2.4e-05</td><td>1.2e-05</td><td>3.1e-05</td><td class="sorting_1">1.55e-05</td><td>dataloader.py:378(__init__)</td></tr><tr role="row" class="odd"><td>24</td><td class="">0.000158</td><td>6.583e-06</td><td>0.000213</td><td class="sorting_1">8.875e-06</td><td>dataloader.py:236(&lt;genexpr&gt;)</td></tr><tr role="row" class="even"><td>1</td><td class="">5e-06</td><td>5e-06</td><td>5e-06</td><td class="sorting_1">5e-06</td><td>dataloader.py:183(DataLoaderIter)</td></tr><tr role="row" class="odd"><td>12</td><td class="">3.1e-05</td><td>2.583e-06</td><td>4.8e-05</td><td class="sorting_1">4e-06</td><td>dataloader.py:331(__del__)</td></tr><tr role="row" class="even"><td>1</td><td class="">1e-06</td><td>1e-06</td><td>1e-06</td><td class="sorting_1">1e-06</td><td>dataloader.py:21(ExceptionWrapper)</td></tr><tr role="row" class="odd"><td>1</td><td class="">1e-06</td><td>1e-06</td><td>1e-06</td><td class="sorting_1">1e-06</td><td>dataloader.py:336(DataLoader)</td></tr></tbody></table></div>

And as a reference I leave here the profiling file (from which come screenshot and details).

[gpu_profiling.zip](https://github.com/YosefLab/scVI-dev/files/1922463/gpu_profiling.zip)

**Comment by jeff-regier** [comment]
What's with `torch._C.mean` taking so long? (122 second?) Does the loss function compute the mean of something?

**Comment by jeff-regier** [comment]
I profiled the code too on aws, and got similar results. Like 80% of run time is in `torch._C.mean`. Got to figure that out.

![snakeviz](https://user-images.githubusercontent.com/8953355/38965678-2b43b6f2-4333-11e8-97b8-7e1e53a3fc1c.png)

**Comment by Edouard360** [comment]
### Batch size

I ran it on Brain Large subsampled 1k, and it seems batch_size doesn't have much influence.

![untitled](https://user-images.githubusercontent.com/15527397/39013016-f52a86de-43ca-11e8-8f7b-2d34074c6fee.png)

I notice though with bigger batch_size, time for ._C.mean slightly decrease, whereas time for  dataloader .get_batch() increases.

### Where does it happen in the code ?

I changed "mean" to "sum", to check that it corresponded exactly in the code to that precise line:

https://github.com/YosefLab/scVI-dev/blob/319047d3b582251f80fd55691f567e2dac1d4862/scvi/scvi.py#L108

And indeed the name of the rectangle in the snakeviz vizualization changed to ._C.sum

### What if I remove some elements ?

Removing the computation of:

https://github.com/YosefLab/scVI-dev/blob/319047d3b582251f80fd55691f567e2dac1d4862/scvi/scvi.py#L95

And replacing the recons loss by 0, 

            reconst_loss = 0

There is a *10 decrease in time for ._C.mean computation.

![screen shot 2018-04-19 at 12 04 33](https://user-images.githubusercontent.com/15527397/39013417-2c7589f8-43cc-11e8-9f2f-7ede0a970442.png)

Then if I leave the computation of the recons_loss but **don't perform the .mean operation on that**, there is still the same (long) time for computation, proving that the computation of log_zinb_wave is actually included under that "._Cmean" rectangle, even though it is not shown in the profiling, and even though this two steps are refering to 2 different parts of the python code...

![screen shot 2018-04-19 at 12 04 43](https://user-images.githubusercontent.com/15527397/39013657-e3fe076c-43cc-11e8-992e-44cfa8f0a7ab.png)

**Comment by jeff-regier** [comment]
Hmmm...any change the  `.mean` operation is getting performed on the CPU? I once saw (what I think was a) Pytorch bug where, because the Python float (in this case, `kl_ponderation`) appeared to the left of the cuda tensor, the cuda tensor was copied to the CPU before the multiply.

**Comment by jeff-regier** [comment]
On second thought, that's probably not it. When I run `make profile` on the aws server, I see GPU usage (measured through `nvidia-smi` exceeding 95%). That suggests nothing is running on the CPU.

**Comment by jeff-regier** [comment]
It could easily be that the profiler is assigning most of runtime of computing `reconst_loss` to the `mean` operation. Perhaps there's some lazy evaluation going on, where `reconst_loss` isn't computed until needed, e.g. for the mean?

Also, the backward pass usually takes about 2x as long as the forward pass. Not clear how the profiler distinguishes btw the backward and forward pass.

**Comment by jeff-regier** [comment]
@Edouard360 So is this still an open issue? You mentioned an indexing problem---is that resolved?

**Comment by jeff-regier** [comment]
Maybe we should start using dense `X` again---I hadn't realized sparse matrix indexing was going to be such a bottleneck. Now I see it's around 50x slower, if done naively using scipy csr_matrix. We could probably get it running 10x faster than it is not (still 5x slower than dense) by fetching a whole minibatch at once. But just converting `X` to dense is probably easier.

For large datasets, we should still use sparse matrices to load the data, and to filter out rare genes before converting it a dense matrix.
